### DISCUSSION TITLE
release vs debug

### DISCUSSION THREAD

**Discussion by tve** [Moddable-OpenSource/moddable#1247] [Q&A]
I'm trying to get a base host app put together that can act as a supervisor for mods that are loaded over the network (primarily mqtt). I'm trying to figure out what permutations of host/mod and workers makes the most sense. This raises some questions:
- what is the overhead of debug builds over release builds of no debugger is connected? Is it just extra data for the debug info? Is there a lot of run-time overhead checking stuff?
- is there any permutation of mod & worker such that a mod or worker can have debug info and the host/main is in release mode? (A worker is a separate VM, so it could act differently?)
- what is the benefit of release mode over debug mode other than making reverse-engineering more difficult and reducing the flash size?

**Comment by phoddie** [comment]
Your project sounds pretty interesting. Delivering mods over MQTT is cool.

> what is the overhead of debug builds over release builds of no debugger is connected? Is it just extra data for the debug info? Is there a lot of run-time overhead checking stuff?

There are two parts to this. There's the build of the host (native Moddable SDK) and the build of the mod. A debug host has the debugger support enabled and some additional bookkeeping for instrumentation. If the debugger is inactive, most of that is pretty light at runtime.

When you build the mod (or any other JavaScript) with debugging enabled, XS includes opcodes to know the current source file and line number. That takes space and some time in execution. If you use a release build of the mod on a debug build, all that overhead goes away.

> is there any permutation of mod & worker such that a mod or worker can have debug info and the host/main is in release mode? (A worker is a separate VM, so it could act differently?)

For RAM & ROM efficiency, the root VM and all workers are cloned from the same preloaded image  use the same Moddable SDK / XS build. So, they are effectively the same. But, as noted above, you could build the JavaScript of the host and mod with difference debug / release settings.

> what is the benefit of release mode over debug mode other than making reverse-engineering more difficult and reducing the flash size?

Smaller firmware image (all debugging support is compiled out) and faster.

Since your mod is coming from the network, you don't need the debugger bridge to install them (which is what mcrun uses). So, a release build should be enough and will be lighter in all respects. But, if you do not have significant performance or memory considerations in your project, a debug build probably won't get in your way.

**Comment by tve** [comment]
So if I understand correct debug host with release mod works but release host with debug mod doesn't due to missing debugger support? Or will it run but just not provide debug info?

Remind me: loading multiple mods is possible in principle but the tools don't support that?

**Reply by phoddie** [reply]
> So if I understand correct debug host with release mod works but release host with debug mod doesn't due to missing debugger support? Or will it run but just not provide debug info?

Debug host with release mod works (and scripts will run a bit faster since they don't have debugging opcodes). Release host with debug mod works (of course you won't be able to debug).

> Remind me: loading multiple mods is possible in principle but the tools don't support that?

It is possible and it works. Two important details:

1. `mcrun` only knows how to install a single mod. With multiple mods you have to manage installing, uninstalling, and loading them yourself.
2. There's no public documentation or examples on multiple mods yet. We do have a private repo for it with a couple of working examples. If you'd like to try it, drop me an email and I'll get you added.
